Add Login and Vote logic as participant

andreslucena opened the pull request:
This PR adds the Login and Voting logic as participant. 

Some notes:

*  As the Census check and Voting is tangled, I prefered to do it in the same PR to save us from added complexity trying to separate both of them 
* Even though I tried to keep it close to `decidim-elections`, there are mainly two differences on how we tackle these issues: 
  * About the login: in Elections the form is checked server-side; here we do it 100% client-side, the only thing that the `LoginForm` does is just to present the form without any kind of validation. 
  * About the code structure, we change the [Change JavaScript logic of window.Decidim.setupVoteComponent](https://github.com/decidim-vocdoni/decidim-module-vocdoni/commit/176b0eb94c5c93d216fd0c726c64265a106d749e). It's fully explained in that commit message. 
* We also have other cosmetic changes regarding the messages (for instance, instead of talking about Encryption, we talk about Submitting the vote, as with the current configuration we're actually not encrypting anything yet, see [commit](https://github.com/decidim-vocdoni/decidim-module-vocdoni/commit/ca70906f855c4154b040e10ddcdaaa650e35ae2a).

Fix #12 and #13 

## Screenshots

![image](https://user-images.githubusercontent.com/717367/218047386-8ce78959-8e6c-4f7e-8725-6a44a8669770.png)

![image](https://user-images.githubusercontent.com/717367/218047403-d7493d79-f0a4-4eaa-bb91-e2380ae1126f.png)

![image](https://user-images.githubusercontent.com/717367/218047421-ecdc5c4a-6b0d-47dd-a99d-165ea0b20a8e.png)

![Selection_503](https://user-images.githubusercontent.com/717367/218048439-e09e1b4b-544f-44b9-a518-2639a26d784e.png)

![image](https://user-images.githubusercontent.com/717367/218048484-3e13c31c-987f-4576-8be2-7b15481e42d9.png)

![image](https://user-images.githubusercontent.com/717367/218048501-dbfd3444-7579-4a3c-94f8-6b841dc97225.png)

![image](https://user-images.githubusercontent.com/717367/218048520-7f889397-6c6c-449f-b522-50e858baacd4.png)

  
## Testing


 
For all, first create and setup an Election that starts shortly (10 minutes from the setup at least just to try it out). 

You need to provide a valid census while creating it. This is an example so it's at hand: 

```csv
email;born_at
user1@example.org;2000-01-01
user2@example.org;2001-02-01
user3@example.org;2001-03-01
user4@example.org;2001-04-01
user5@example.org;2001-04-01
user6@example.org;2001-04-01
user7@example.org;2001-04-01
user8@example.org;2001-04-01
```

Then....

### Login form

- [ ] Given that I'm a voter in the census
      When I go to vote during the voting process
      I can verify my data with the one provided by an admin #7 
- [ ] Given that I'm a voter in the census
      With an email address like "admin@example.org"
      When I try to verify with "ADMIN@EXAMPLE.ORG"
      Then the system verifies me correctly
- [ ] Given that I'm a voter in the census
      When I go to verify myself
      Then I see a help text in the "born at" field with the expected format (like "Format: day/month/year")
      And when I click in this field
      Then I have a date field selector

### Voting

- [ ] Given that I'm a voter that has already logged in
       When I finish my vote
       I have a voterId
       and I can check it with the Vocdoni Explorer portal (example https://dev.explorer.vote/verify/#/3c1e8371071f15aa031de5bb1e3a611a206aaa0b868c0fa0921cd0a42c75b018)

microstudi commented on app/controllers/concerns/decidim/vocdoni/has_vote_flow.rb:
```diff
@@ -0,0 +1,32 @@
+      end
+
+      def preview_mode?
+        return @preview_mode if defined?(@preview_mode)
+
+        @preview_mode = !election.published? || !election.started?
```
```suggestion
        @preview_mode ||= !election.published? || !election.started?
```
Can't we just do this?

andreslucena replied:
Yes, nice catch, much cleaner!

This was actually copied from `decidim-elections`: https://github.com/decidim/decidim/blob/84b97f904c72ba2bb9037721e63b4fb3b29fbaa0/decidim-elections/app/controllers/concerns/decidim/elections/has_vote_flow.rb#L26

microstudi commented on app/controllers/concerns/decidim/vocdoni/has_vote_flow.rb:
```diff
@@ -0,0 +1,32 @@
+      end
+
+      def can_preview?
+        return @can_preview if defined?(@can_preview)
+
+        @preview_mode = allowed_to?(:preview, :election, election: election)
```
```suggestion
        @preview_mode ||= allowed_to?(:preview, :election, election: election)
```
same here

microstudi commented on app/packs/src/decidim/vocdoni/voter/new-vote.js:
```diff
@@ -0,0 +1,212 @@
+       *
+       * @return {array} An array with the selected answers
+       */
+      const getFormData = (formData) => {
+        /* eslint-disable no-invalid-this */
+        return formData.map(function() {
+          const answerValue = this.value;
+          return Number(answerValue.split("-")[1]);
+        }).get();
+        /* eslint-enable no-invalid-this */
+      };
+      const formData = getFormData($voteWrapper.find(".answer_input:checked"));
```
```suggestion
      const $checks = $voteWrapper.find(".answer_input:checked")
      const formData = $checks.map((_idx, el) => el.value.split("-")[1]}).get());
```
I find the use of "this" confuse in this context (as is comming from jquery but not very clear), maybe we could rewrite it to something like this

andreslucena replied:
Much clean indeed 😍 

I need to fix a couple details: 

1. There's a leftover closing curly brace there ("}")
>       const formData = $checks.map((_idx, el) => el.value.split("-")[1]**}**).get());

2. we need to convert to Number as that's what the SDK expects. 

I made the changes on your suggestion so I can just accept it, but great refactor!

microstudi commented on lib/decidim/vocdoni/component.rb:
```diff
@@ -6,6 +6,7 @@
   component.engine = Decidim::Vocdoni::Engine
   component.admin_engine = Decidim::Vocdoni::AdminEngine
   component.icon = "media/images/decidim_vocdoni.svg"
+  component.stylesheet = "decidim/vocdoni/vocdoni"
```
This introduces the stylesheet in every page along Decidim. Maybe we should make sure to namespace any potential CSS class that can have collisions with other with the same name in places unrelated. For instance, I'm thinking on the ".accordion" class.

andreslucena replied:
Oh I see what you mean. I missed that.

Done at aad92441ee9fb2fe024bb242da7f8634655dc387

microstudi reviewed (approved):
Excellent job @andreslucena 

I've left a couple of comments, see if they are relevant.

Notes for future PRs:
- [ ] We should collect any possible error from javascript and inform the user, I got a blank page after voting and the console showed this: ![Screenshot from 2023-02-22 18-17-16](https://user-images.githubusercontent.com/1401520/220710151-d0e69cd7-f14f-416a-9020-1f1f5fd57361.png)
- [ ] In the admin side, it would be very nice to guide the admin step by step, a sort of "Next step" button to avoid going back and press the option required to complete all the steps.

andreslucena commented:
> [ ] We should collect any possible error from javascript and inform the user, I got a blank page after voting and the console showed this:

I agree. This was a bug with the selector for showing the error message. It was solved on https://github.com/decidim-vocdoni/decidim-module-vocdoni/pull/25/commits/77f86893478832bb375c28959805c612d2243075 

> [ ] In the admin side, it would be very nice to guide the admin step by step, a sort of "Next step" button to avoid going back and press the option required to complete all the steps.

We discussed this with @carolromero last week. A couple comments regarding what we've talked: 

1. I tried to stay close to the `decidim-elections` flow, so when an Admin understands one its easy to understand the other. Also, much of the code is copied from this module, with the idea of not closing the door to proposing the introduction of this evoting provider to that module. 
2. I tried to stay close to the "decidim admin way", meaning that the controllers are separated even when they don't make much sense. I agree that this is an overhead for some actions, that make some features/flows not discoverable by admins (for instance all the Meetings forms and features in the admin). I think this should be handled with the redesign, and then we need to implement the new way here. 
3. As a way of mitigating this usability problem, we've changed two things: 
3.1. The order of actions in the admins' elections index page, so its easy to arrive to the Steps page: https://github.com/decidim-vocdoni/decidim-module-vocdoni/pull/36/commits/bb4aabf9e3c855390ec2ec00b28f80aa78416b07 
3.2. We added a "Fix it" link when one check fails in this Steps page, so its easy to navigate: https://github.com/decidim-vocdoni/decidim-module-vocdoni/commit/4af79883740f27c403f62bd97af840fcb7f4114a

andreslucena commented:
@microstudi as you've approved the PR before and I also applied your suggestions, I'll go ahead and merge this one so we can follow with the others. Thanks for your feedback!
